### ISSUE TITLE
[Linux] Use more specific & more integrated alternatives for status icon

### ISSUE THREAD

**Issue by csoriano1618** [telegramdesktop/tdesktop#3830]
Hello,

Carlos from Linux & GNOME here. We are trying to approach important apps using status icons to provide better alternatives that are more integrated with the system. Status icons have been deprecated for some time and we plan to remove the systray for GNOME 3.26.

Currently Telegram uses status icon for:
1. Notify of incoming messages
2. Quit application
3. Disable notifications
4. Minimize to tray
5. Open app

For 1, the proper approach is to use native notifications. Telegram has a setting for this, but it's off by default. The solution would be to enable this on by default, so the user benefits from the notification integration.
This will allow the user to see whether the user has any pending messages in a permanent place, and also disable notifications if desired in a grained way.
I filed issue #3831 separately to discuss the notification implementation. Documentation for native notifications can be found [here](https://developer.gnome.org/hig/stable/notifications.html.en).
For 2, the recommended approach is to ask the user if the application should run in the background the first time it click the close button, then enable a setting with "run in the background". The user can deactivate it anytime if changes user's mind.
An example would be Polari, and IRC client. The dialog looks like:
![run in background](https://user-images.githubusercontent.com/1755514/29708327-8cdbbe88-8988-11e7-8b7c-a8e175e5dcf6.png)

For 3, would be fixed with 1, and will use the system wide settings and integration for notifications, instead of something for every app in the system.
For 4, won't be needed if 2 is implemented.
For 5, it happens when a notification is clicked, or if the app is run again from the recent app list or from any place where apps can be searched & run

Something to note here is that these solutions are cross-desktop, so it should work for any Linux distribution.

Let me know what do you think and feel free to ask any question!

Thanks

**Comment by stek29** [comment]
> Quit application

Ctrl+Q

> Disable notifications

It's in settings

**Comment by fbruetting** [comment]
@csoriano1618 I’m wondering, what the GNOME view of the message aggregation indicator of Unity is? Since messages are usually more important than other (non-permanent) notifications, one can separate urgent cases from the minor ones. Also this would target a whole class of use cases (mails, instant messages, IRC, Twitter, etc. pp.). The benefits I see are:

• Not as crowded as a list of mixed notifications about all kinds of things in the notification center.
• For instant messaging services, people usually respond very quickly – while most people (I know of) just look sparsely into the notification center. Missing instant messaging notifications prevents messaging being instant.
• One should be able to mark notification sources as inactive. E.g. receiving E-Mails should not light up the message indicator, because most mails are not that important. You should still see the number of unread mails of every mail account when you click the indicator, but this way you won’t be compelled clicking it. Received instant messages should indeed light it up like it’s done on everybody’s mobile. One then would be able seeing all unread messages of all sorts of messaging sources all together in a clean and unified way.
• I think that communication is so important, that it deserves it’s own icon in the top panel. Maybe this aggregation could be implemented into the notification center, instead of making up a separate indicator, but communication overview should be separated from other notifications.
• This approach could solve all mentioned issues and introduce more coherence into desktop environments.

**Comment by diazbastian** [comment]
@csoriano1618 I would also like to see native notifications by default, but currently Telegram notifications offer better features. On the other hand there is no native solution for voice calls (eg something similar to the implementation of MPRIS in Gnome, where you can accept/deny a call and view its status persistently without go to the app window) . The above example would also work for any other voice and video application (Ring, Skype, Viber, Gajim, Jitsi, etc.).

_Note: You may be interested in the ticket_ [279](https://github.com/transmission/transmission/issues/279) _of transmission_.

**Comment by csoriano1618** [comment]
Thanks for the feedback and questions so far. I'll answer them soon, I'm evaluation the questions with our designers to provide accurate answers, sorry for taking long :)

**Comment by Aokromes** [comment]
#2191 ?

**Comment by aruiz** [comment]
Hey @csoriano1618, any progress on this? :-)

**Comment by csoriano1618** [comment]
Hey,

Sorry for the late answer. Indeed, Telegram has a good point that native notifications doesn't implement an easy way to reply inline. GNOME actually wants that too, so it's a matter of working on extending the protocol.

**Comment by cold-distance** [comment]
Hi, I saw this bug report so I will take it to make a suggestion.

I use Fedora with GNOME with vanilla experience now, so I don't have status icons or systray.

It would be wonderful if Telegram Desktop could be closed only with the close button. The only thing needed is to have an option in settings to disable systray/status icon support at least in Linux.

**Comment by jhasse** [comment]
There is.

**Comment by fbruetting** [comment]
@csoriano1618 What about your answers? :P

> Thanks for the feedback and questions so far. I'll answer them soon, I'm evaluation the questions with our designers and other developers to provide accurate answers, sorry for taking long :)

**Comment by csoriano1618** [comment]
Hey,

I was rereading now the comments. I think I answered the most important one, which is what about inline reply. Last time we discussed it we indeed wanted that, and we would like to extend the protocol to include it if possible.

For the design of the "notification center" in GNOME Shell, that's probably not very important here, since it's up to the DE how to interpret those notifications. That probably isn't much of a requirement either, since Telegram doesn't implement the proposed design currently either.

> On the other hand there is no native solution for voice calls

Wouldn't a notification with two actions (accept, deny) be the solution for this?

Is there anything I'm missing? I'll be glad to get an answer for those.

**Comment by fbruetting** [comment]
Ah, I thought you can speak for Gnome about a messaging center concept. 😄

**Comment by diazbastian** [comment]
> Wouldn't a notification with two actions (accept, deny) be the solution for this?

@csoriano1618 
Maybe, but I was referring to an implementation more similar to that found on mobile devices (persistent notification).
For example in Android, if we don't answer immediately or we are occupying another app, you can accept a call from a notification or the notification center, the same to stop or cancel a call. This also applies to other apps that include calls and allow you to do this without having to bring the app to focus. The most common options are accept/decline /mute microphone.

**Comment by csoriano1618** [comment]
> This also applies to other apps that include calls and allow you to do this without having to bring the app to focus.

Yeah that makes sense, there has been some discussion about putting actions on the notification center.

I'm not sure this is in scope of this issue tho... the native notification issue is at https://github.com/telegramdesktop/tdesktop/issues/3831. But also, Telegram notifications cannot do this kind of thing right now either, it would need collaboration with the DE, and indeed it would be amazing to improve that. With this issue I'm more focusing in covering what the current Telegram notifications provide and make sure those are ok, so Telegram can transition to native notifications without big issues.

**Comment by pioruns** [comment]
Two years and still not fixed? What is going on?

**Comment by GabVenturato** [comment]
@Aokromes @csoriano1618 I think the problem now is that in Gnome (I am referring to my experience in Fedora 30, you can read more in che closed issue referred above) I have no option to have telegram closed when the close button is pressed! If I want to close telegram I have to kill the process by CLI. Is there a fix for this? Because it seems to me that the situation is a bit neglected.

**Comment by jhasse** [comment]
@GabVenturato "Settings" -> "Advanced" -> Uncheck "Show tray icon"

**Comment by cold-distance** [comment]
> @GabVenturato "Settings" -> "Advanced" -> Uncheck "Show tray icon"
> 
> edit: after reading #6412 I understand your problem: The option isn't there for you in GNOME.

I just checked and I see that the problem persists. It's very annoying.

**Comment by GabVenturato** [comment]
@jhasse yeah, that's exactly the problem

**Comment by DarkFenX** [comment]
#6692 was created by me. Needless to say, I am annoyed by lack of tray icon.

I think suggested approach does not follow the way I am using telegram. I use it at work as IM for personal purposes. So, I keep it open, with tray icon shown (via TopIcons). When I focus on my task - I do not want any kind of popups to distract me (let alone fellow coworkers looking at content of notifications). However, when I have a moment, I take a look at tray and see if there're any unread discord/telegram notifications.

And it worked just fine, until today.

How should I approach my use-case within "new" gnome paradigm?

**Comment by ilya-fedin** [comment]
It seems no one stopped stale bot from closing the issue despite it's apparently not fixed as duplicates of this issue are being still created... Sadly, there's still no progress in solving this, I've tried multiple approaches, but all of them lead into problems due to the lack of APIs.

Since then, native notifications became the default based on supported capabilities of the notification daemon, the currently required capabilities (body, actions, inline-reply, inhibitions) practically mean native notifications are default only on KDE as it's the only server implementation supporting all of them. Here's what I tried and what could help to make native notifications the default on GNOME:
1. The most obvious solution of using the `sound` capability and letting the server playing the sounds runs into the fact notifications daemons accept only WAVs while Telegram supports mp3 and wav files which are being downloaded from the cloud, so they have to be re-encoded on the fly and sadly there's no one who have both time and knowledge at the same time in tdesktop project to implement a re-encoder. This hasn't happen for 6 years and I'm highly doubt this can ever happen unless GNOME project provides help with code.
2. I've tried to make the use of the `sound-name` property to play the sound provided by the system, but as it happened the spec guarantees only support for `sound-file` and `suppress-sound` hints with that capability which turns this into no-go. I've [reported that at xdg-specs](https://gitlab.freedesktop.org/xdg/xdg-specs/-/issues/76) two years ago and even [created a MR](https://gitlab.freedesktop.org/xdg/xdg-specs/-/merge_requests/70) both of which are ignored. Maybe GNOME can get attention to those?
3. The most simplest solution for tdesktop would be for GNOME to support both `inhibitions` and `inline-reply` as that would mean zero code changes from tdesktop side.

The problem with running in background is still valid as well and there's no permission from the designers to add such a dialog AFAIK. Possible solution is to use the macOS code path where it's assumed it's native to the system to run in background and it will automatically detect the app is still running and provide a way to close it (e.g. with an icon in the dock as macOS does). The problem here is we need an API to know that the DE prefers apps to run in background, something like a property on org.freedesktop.portal.Settings would fit pretty nice.

@csoriano1618 are you still interested in solving this? Can you provide help with code or standardizing the APIs and having them implemented on GNOME side?

**Comment by ilya-fedin** [comment]
Ah, forgot to mention that I also attempted to use [GNotifcation](https://docs.gtk.org/gio/class.Notification.html). I thought it would be nice to use it to support the portal notifications and I saw it also supports some proprietary GNOME spec, so I thought maybe it's something GNOME notifications use to provide the good UX with sounds. After getting notifications implemented on top of GNotification, I understood that it doesn't support sounds at all with [frustrating finding that GNOME doesn't want to support sounds](https://bugzilla.gnome.org/show_bug.cgi?id=750931#c6), but it does via the freedesktop spec what makes GNotification an outsider in comparison to the freedesktop spec (and there's a hidden feature that makes the use of sound-name when the play sound tdesktop option is off), as a result GNotification is used only in flatpak and only when the freedesktop protocol is not available. As GNotification doesn't provide a way to detect whether there's support whether there's any protocol support, it's not possible to use it by default for all. It also doesn't support:
1. Images via path for the portal
2. Images via bytes for the freedesktop spec
What means it's not a good API to use by default, as there's always a chance someone won't see the avatar as it's not possible to set bytes and path at the same time. Btw, while glib can't use gdk-pixbuf, I believe it can at least create a temporary file from the bytes for the freedesktop spec.

I don't really understand how GNOME messengers provide good UX with notifications, do GNOME users like when there's no sound on incoming message? As I'm highly doubt tdesktop userbase, even the one that is using GNOME, will appreciate some blacklist to not to play sound on GNOME.

**Comment by ilya-fedin** [comment]
@AsciiWolf please read the previous posts in the thread, we need APIs from GNOME side (& to be standardized) in order to run in background without window. However, there's already "Close to taskbar" option that lets Telegram run in background using cross-platform APIs to hide the window instead of destroying it.

**Comment by AsciiWolf** [comment]
What I am talking about is support for Telegram to run on background on systems without AppIndicator / KStatusNotifierItem support (or did you mean other APIs?). For example Steam from Valve seems to run fine on background even on a stock GNOME and when subsequently launching the Steam client, window for the active session that is running on background gets opened. At least this is how it worked the last time I tried it on a stock GNOME system. Telegram is the main reason I am still having the (buggy) AppIndicator extension enabled on all my GNOME systems.

**Comment by ilya-fedin** [comment]
@AsciiWolf Electron doesn't have API to check tray presence, so I assume they have a bug of running in background with no way for the user to quit the application (e.g. some lightweight window manager environment without tray support and without desktop actions support). Telegram being a Qt application can act correctly in that case and acts. I don't think we should leave those users without a way to quit in the sake of GNOME users convenience (implementing a feature request while adding a bug). Actually, I've explained that all in older messages in the thread, please read them and don't make me to repeat all that again & again.

**Comment by lucacarpizzo836** [comment]
I think there's some misunderstanding going on here. GNOME 44 has added support for background apps, what's preventing the implementation of that feature into the app?

**Comment by lucacarpizzo836** [comment]
@ilya-fedin I think there's some misunderstanding going on here. GNOME 44 has added support for background apps, what's preventing the implementation of that feature into the app?

**Comment by jtojnar** [comment]
@lucacarpizzo836 that is currently only implemented for Flatpak apps, see https://gitlab.gnome.org/GNOME/gnome-shell/-/issues/3699

**Comment by AsciiWolf** [comment]
The plan as far as I know is to extend the background portal to also support non-sandboxed apps in the future. Anyway, as I already mentioned in my previous comment, Telegram does not need to support the background portal API directly, it just needs to be able to continue running on background (even without the AppIndicator icon) after its main window gets closed + to be able to restore the background session when subsequently launching Telegram.

**Comment by ilya-fedin** [comment]
> what's preventing the implementation of that feature into the app?

As I already said, we need API to know whether the environment prefers applications running in background to closing them, as Telegram designers don't give a permission to implement a dialogue like proposed here in first post, so we need a solution adding zero new UI elements to tdesktop. Such API is the only way I can see.

**Comment by ilya-fedin** [comment]
The API has no "Supported" method and I have the interface on Plasma (while I know for sure it has no UI for background apps) so the presence of the interface itself doesn't say anything about support either. So this is not the API I was describing.

Basically, we need a way to ensure the user sees the quit button, so it may be not an API to check whether the environment prefers applications running in background but an API to check whether Desktop Actions are supported by the environment. But I don't see such an API either.

**Comment by AsciiWolf** [comment]
It is also supported on Plasma, see: https://github.com/KDE/xdg-desktop-portal-kde/blob/master/src/background.cpp

I may be wrong, but I think that it also is somehow supported on non-GNOME/non-KDE environments by directly using the xdg-desktop-portal (not relying on xdg-desktop-portal-gnome or xdg-desktop-portal-kde).

**Comment by ilya-fedin** [comment]
@AsciiWolf I don't see any indicator that application is running in background, not sure we can say it's really supported

**Comment by ilya-fedin** [comment]
I'm doubt the presence of background portal means 'this environment prefers the application to run in background and is obligated to provide a background indicator, desktop actions or other means to let the user quit the application'.

**Comment by AsciiWolf** [comment]
There should be a notification. See for example this thread: https://www.reddit.com/r/kde/comments/11ul75n/is_there_a_way_to_disable_this_running_in_the/

**Comment by ilya-fedin** [comment]
The notification won't allow me quit the application.

**Comment by AsciiWolf** [comment]
Clicking "Find out more" should allow you to force quit the app. I agree that the KDE implementation may not be ideal (but will hopefully improve over time), but it works. And using this background portal (at least optionally) would, in my opinion, still be much better than shipping the Telegram app in a broken state to all stock GNOME users.

**Comment by ilya-fedin** [comment]
I looked at the portal documentation, it's just to request background permission in flatpak and has nothing to do with

> this environment prefers applications to run in background and is obligated to provide a background indicator, desktop actions or other means to let the user quit the application

**Comment by AsciiWolf** [comment]
I don't use Plasma so I am not sure what the implementation there exactly looks like.

**Comment by ilya-fedin** [comment]
> than shipping the Telegram app in a broken state to all stock GNOME users.

The key problem is the lack of contributors. I'm the only one Linux user contributing to tdesktop and I'm a Plasma user, I have no idea how applications should work on GNOME. For Telegram to be not broken on GNOME, someone from the GNOME community with the knowledge how applications should work there has to step up and contribute fixes for Telegram for it to be no more broken. There's even no contributor to do [that](https://github.com/telegramdesktop/tdesktop/issues/16062#issuecomment-816868341).

**Comment by ilya-fedin** [comment]
> I don't use Plasma so I am not sure what the implementation there exactly looks like.

Not sure what you mean. I talk about documentation provided by the links you provided
> org.freedesktop.portal.Background — Portal for requesting autostart and background activity

**Comment by ilya-fedin** [comment]
> Clicking "Find out more" should allow you to force quit the app. I agree that the KDE implementation may not be ideal (but will hopefully improve over time), but it works. And using this background portal (at least optionally) would, in my opinion, still be much better than shipping the Telegram app in a broken state to all stock GNOME users.

It provides the dialog to grant or reject the permission, that's not what I'm talking about. I'm talking about quitting the application while already running in background cause the app UI has no way to quit in the UI (and no one would allow to add it).

**Comment by ilya-fedin** [comment]
Another way out of this is push notifications so tdesktop doesn't have to run in background at all. But there seem to be no news about push notifications API either...

**Comment by AsciiWolf** [comment]
> Not sure what you mean. I talk about documentation provided by the links you provided

As far as I know, the Background portal (and/or its implementation in xdg-desktop-portal-gnome) was extended when the new "Background Apps" dialog was implemented in GNOME 44. See for example [this ticket](https://github.com/flatpak/xdg-desktop-portal/issues/899).

**Comment by ilya-fedin** [comment]
If I understand your link right, they just added a way for the portal implementation to know which sandboxed applications are running. This has nothing to do with

> this environment prefers applications to run in background and is obligated to provide a background indicator, desktop actions or other means to let the user quit the application

**Comment by AsciiWolf** [comment]
They added a way for the portal implementation to know which sandboxed applications are running **on background** and communicating this to a host system via D-Bus.

I am however not sure whether the Background portal can detect if the "environment prefers applications to run in background" and can pass this information to the application via API.

**Comment by ilya-fedin** [comment]
> They added a way for the portal implementation to know which sandboxed applications are running **on background** and communicating this to a host system via D-Bus.

Yes. That's the problem: how GNOME uses this information is just implementation detail, we can't rely on implementation detail.

**Comment by ilya-fedin** [comment]
I've got an idea but not sure how good it is: check title controls layout and if it has no minimize button, assume minimization is not supported and run in background.

**Comment by AsciiWolf** [comment]
Sounds like an interesting idea. :-)

**Comment by AsciiWolf** [comment]
I have thought about the solution suggested by ilya-fedin and I think that it is probably the best possible solution to this problem at the moment. If it could be implemented, it would be great and would solve the problem for most GNOME users.

Just make sure that re-opening the Telegram app from the applications menu will open window for the already running session instead of opening a new session. :-) But I think that this already works correctly, at least when using SNI/systray (AppIndicator extension for GNOME Shell) so it will hopefully work the same even without the extension.

**Comment by ilya-fedin** [comment]
It's already implemented in 4.9.6

**Comment by AsciiWolf** [comment]
Great work, thanks!

**Comment by AsciiWolf** [comment]
I have just tested it on my Fedora 38 system with GNOME 44 and everything seems to work great! Telegram now correctly stays on background after closing its window, re-opening the Telegram app correctly opens the running session and Telegram session can be properly closed either by clicking "Quit Telegram" in GNOME Shell Overview or by closing the app from the GNOME Background Apps menu. Thanks a lot for resolving this issue for us GNOME users, ilya-fedin! :-)

**Comment by BachoSeven** [comment]
@ilya-fedin Why was this implemented so that the default on environments with no minimization is to run in background, but with no option to tweak so?

(I assume so by looking at https://github.com/telegramdesktop/tdesktop/pull/26777 and by checking in the settings, where I couldn't find any option)

**Comment by ilya-fedin** [comment]
Because there's still no contributor to implement the option.

**Comment by ilya-fedin** [comment]
So I took the simplest solution - do in such environments like on macOS re-using already existing code path

**Comment by BachoSeven** [comment]
So if I'm on a Linux environment with no system tray or minimization I can't prevent Telegram from going into background on close?

I remember the option being there in the somewhat-recent past(?)

**Comment by AsciiWolf** [comment]
Running on background used to work only on systems with AppIndicator support and you could disable the AppIndicator (tray icon) support in Telegram settings which probably prevented Telegram from running on background. To be honest, I don't see this as a big deal since most users probably want their IM client to run on background. And you can always click "Quit Telegram" instead of just closing its window - it's just about the muscle memory. :-)

**Comment by ilya-fedin** [comment]
> So if I'm on a Linux environment with no system tray or minimization I can't prevent Telegram from going into background on close?

You just need to remember to use the Quit Telegram menu item (desktop action) instead of close button and Ctrl+Q instead of Alt+F4.

**Comment by eternal-sorrow** [comment]
Why would anyone want that though? To have an application running in the background unchecked without any indication of it?

**Comment by AsciiWolf** [comment]
Why not? Most users use their IM clients this way. They just want them to run on background so they are online, can receive notifications etc. Telegram Desktop now works 100% correct on stock GNOME and we as users can't be happier. ;-)

**Comment by eternal-sorrow** [comment]
You can minimize the window or move it to another workspace if it somehow hinders you. And still have an indication the the app is still running. As it is now, there is no way of knowing if the app is running in the background or maybe it has crashed for example.

**Comment by AsciiWolf** [comment]
> You can minimize the window or move it to another workspace if it somehow hinders you. And still have an indication if the app is still running.

No, this is definitely not how most users (including me) want to use Telegram and it was bothering us for years.

Also, you are not entirely correct that there is "no way of knowing if the app is running in the background". If you use a Flatpak version of Telegram Desktop, it shows in the "Background Apps" list that is in stock GNOME since 44.

**Comment by eternal-sorrow** [comment]
Not everyone uses GNOME though.

**Comment by AsciiWolf** [comment]
Fair point, however how many desktop environments except GNOME have no minimize button in their title controls layout?

**Comment by eternal-sorrow** [comment]
>have no minimize button in their title controls layout?

That is your problem? You don't know how to restore the minimize button in GNOME?

>how many desktop environments except GNOME have no minimize button in their title controls layout?

My environment has no title and no window control buttons whatsoever. My problem is that the process `telegram-desktop` continues to persist after I pressed the close window shortcut. I wanted to restart the telegram app after I updated It but was unable to without using a `killall telegram-desktop` command.

**Comment by eternal-sorrow** [comment]
>I am not sure how tdesktop checks the layout there

Does it check anything at all? I doubt it.

**Comment by ilya-fedin** [comment]
You may have GNOME settings installed, check with dconf-editor

**Comment by eternal-sorrow** [comment]
>You may have GNOME settings installed, check with dconf-editor

What exactly should I check?

**Comment by ilya-fedin** [comment]
I write from smartphone, so I can't check exactly. Find decoration-layout, button-layout or something like that

**Comment by eternal-sorrow** [comment]
Wait, so you're saying that Telegram looks into GNOME settings to determine if minimization supported by the environment and that's how it decides if it needs to run in the background or not?

This is only relevant in GNOME, you know.

**Comment by ilya-fedin** [comment]
GNOME's key is the only key that exists for this as far as I'm aware. It has a wide support as gtk reads only it so most DEs including KDE set this key.

**Comment by eternal-sorrow** [comment]
As I said, some environments may not have the window controls at all. Others may have it, but ignore the GNOME settings. Some may not even support a concept of window minimization. Just add a preference in Telegram settings, don't rely solely on an obscure setting relevant only to a specific environment.

**Comment by ilya-fedin** [comment]
Forget about window control preference, designers won't allow that. So either the Linux community standardizes some standard cross-DE key or it will work as it works.

**Comment by eternal-sorrow** [comment]
>designers won't allow that

And why is that?

**Comment by eternal-sorrow** [comment]
For everyone like me who encounters this problem, workaround is to set the dconf key `/org/gnome/desktop/wm/preferences/button-layout` to something that contains the `minimize` keyword, for example `appmenu:minimize,close`.

**Comment by ilya-fedin** [comment]
> And why is that?

IDK actually. I remember I talked to @john-preston about adding such a setting when porting the Telegram titlebar to Linux, he disallowed me to add such a setting. So I had to implement on top of what I want, what I wanted back then was the hardcoded Windows layout. The reading of the GNOME key was added a bit later, first by using gtk and later rewritten to reading portal + xsettings when portal communication code was written and xsettings implementation was stolen from Qt.
